# 项目：整理Netflix电影演员评分数据

## 分析目标

此数据分析的目的是，整理不同流派影视作品，比如喜剧片、动作片、科幻片中，各演员出演作品的平均IMDB评分，从而挖掘出各个流派中的高评分作品演员。

本实战项目的目的在于练习整理数据，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了截止至2022年7月美国地区可观看的所有Netflix电视剧及电影数据。数据集包含两个数据表：`titles.csv`和`credits.csv`。

`titles.csv`包含电影及电视剧相关信息，包括影视作品ID、标题、类型、描述、流派、IMDB（一个国外的在线评分网站）评分，等等。`credits.csv`包含超过7万名出现在Netflix影视作品的导演及演员信息，包括名字、影视作品ID、人物名、演职员类型（导演/演员）等。

`titles.csv`每列的含义如下：
- id：影视作品ID。
- title：影视作品标题。
- show_type：作品类型，电视节目或电影。
- description：简短描述。
- release_year：发布年份。
- age_certification：适龄认证。
- runtime：每集电视剧或电影的长度。
- genres：流派类型列表。
- production_countries：出品国家列表。
- seasons：如果是电视剧，则是季数。
- imdb_id：IMDB的ID。
- imdb_score：IMDB的评分。
- imdb_votes：IMDB的投票数。
- tmdb_popularity：TMDB的流行度。
- tmdb_score：TMDB的评分。

`credits.csv`每列的含义如下：
- person_ID：演职员ID。
- id：参与的影视作品ID。
- name：姓名。
- character_name：角色姓名。
- role：演职员类型，演员或导演。

## 评估和清洗数据

In [1]:
import pandas as pd

In [2]:
import numpy as np

In [3]:
df_titles = pd.read_csv('titles.csv')
df_credits = pd.read_csv('credits.csv')

### 查看数据结构的整齐度

发现`genres`列的单元格的内容不是一个值，应当根据该列的数据类型将其拆分成多行，使得每个单元格都是一个值。

In [4]:
df_titles['genres'] = df_titles['genres'].apply(lambda x : eval(x))

In [5]:
df_titles = df_titles.explode('genres')

In [6]:
df_titles

id                                      title   type  \
0      ts300399        Five Came Back: The Reference Films   SHOW   
1       tm84618                                Taxi Driver  MOVIE   
1       tm84618                                Taxi Driver  MOVIE   
2      tm154986                                Deliverance  MOVIE   
2      tm154986                                Deliverance  MOVIE   
...         ...                                        ...    ...   
5847  tm1059008                                    Lokillo  MOVIE   
5848  tm1035612  Dad Stop Embarrassing Me - The Afterparty  MOVIE   
5849   ts271048         Mighty Little Bheem: Kite Festival   SHOW   
5849   ts271048         Mighty Little Bheem: Kite Festival   SHOW   
5849   ts271048         Mighty Little Bheem: Kite Festival   SHOW   

                                            description  release_year  \
0     This collection includes 12 World War II-era p...          1945   
1     A mentally unstable Vietnam War veteran works ...          1976   
1     A mentally unstable Vietnam War veteran works ...          1976   
2     Intent on seeing the Cahulawassee River before...          1972   
2     Intent on seeing the Cahulawassee River before...          1972   
...                                                 ...           ...   
5847  A controversial TV host and comedian who has b...          2021   
5848  Jamie Foxx, David Alan Grier and more from the...          2021   
5849  With winter behind them, Bheem and his townspe...          2021   
5849  With winter behind them, Bheem and his townspe...          2021   
5849  With winter behind them, Bheem and his townspe...          2021   

     age_certification  runtime         genres production_countries  seasons  \
0                TV-MA       51  documentation               ['US']      1.0   
1                    R      114          drama               ['US']      NaN   
1                    R      114          crime               ['US']      NaN   
2                    R      109          drama               ['US']      NaN   
2                    R      109         action               ['US']      NaN   
...                ...      ...            ...                  ...      ...   
5847               NaN       90         comedy               ['CO']      NaN   
5848             PG-13       37            NaN               ['US']      NaN   
5849               NaN        7         family                   []      1.0   
5849               NaN        7      animation                   []      1.0   
5849               NaN        7         comedy                   []      1.0   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0            NaN         NaN         NaN            0.600         NaN  
1      tt0075314         8.2    808582.0           40.965       8.179  
1      tt0075314         8.2    808582.0           40.965       8.179  
2      tt0068473         7.7    107673.0           10.010       7.300  
2      tt0068473         7.7    107673.0           10.010       7.300  
...          ...         ...         ...              ...         ...  
5847  tt14585902         3.8        68.0           26.005       6.300  
5848         NaN         NaN         NaN            1.296      10.000  
5849  tt13711094         7.8        18.0            2.289      10.000  
5849  tt13711094         7.8        18.0            2.289      10.000  
5849  tt13711094         7.8        18.0            2.289      10.000  

[15147 rows x 15 columns]

In [7]:
df_credits

person_id         id                name                character  \
0           3748    tm84618      Robert De Niro            Travis Bickle   
1          14658    tm84618        Jodie Foster            Iris Steensma   
2           7064    tm84618       Albert Brooks                      Tom   
3           3739    tm84618       Harvey Keitel  Matthew 'Sport' Higgins   
4          48933    tm84618     Cybill Shepherd                    Betsy   
...          ...        ...                 ...                      ...   
77796     736339  tm1059008    Adelaida Buscato                María Paz   
77797     399499  tm1059008  Luz Stella Luengas             Karen Bayona   
77798     373198  tm1059008         Inés Prieto                    Fanny   
77799     378132  tm1059008        Isabel Gaona                   Cacica   
77800    1950416  tm1059008      Julian Gaviria                      NaN   

           role  
0         ACTOR  
1         ACTOR  
2         ACTOR  
3         ACTOR  
4         ACTOR  
...         ...  
77796     ACTOR  
77797     ACTOR  
77798     ACTOR  
77799     ACTOR  
77800  DIRECTOR  

[77801 rows x 5 columns]

`df_titles`&`df_credits`两个df的结构整齐度已达标。

### 查看数据内容的干净度

#### df_titles

In [8]:
df_titles.info()

<class 'pandas.core.frame.DataFrame'>
Index: 15147 entries, 0 to 5849
Data columns (total 15 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    15147 non-null  object 
 1   title                 15146 non-null  object 
 2   type                  15147 non-null  object 
 3   description           15125 non-null  object 
 4   release_year          15147 non-null  int64  
 5   age_certification     9298 non-null   object 
 6   runtime               15147 non-null  int64  
 7   genres                15088 non-null  object 
 8   production_countries  15147 non-null  object 
 9   seasons               5923 non-null   float64
 10  imdb_id               14525 non-null  object 
 11  imdb_score            14399 non-null  float64
 12  imdb_votes            14375 non-null  float64
 13  tmdb_popularity       14995 non-null  float64
 14  tmdb_score            14614 non-null  float64
dtypes: float64(5), int64(2), 

该数据集中有15147行观察值：

缺失值：`title`,`description`,`age_certification`,`genres`,`seasons`,`imdb_id`,`imdb_score`,`imdb_votes`,`tmdb_popularity`,`tmdb_score`均有缺失值。

数据类型错误：`release_year`应转为datetime,`production_countries`应转为category,`seasons`应转为int.

因为本次数据分析的目的是，寻找出各个流派中平均作品评分最高的演员，所以`genres`,`imdb_score`是重要的变量，如若缺失，应删除。

In [9]:
df_titles.dropna(subset = ['genres'],inplace = True)

In [10]:
df_titles.dropna(subset = ['imdb_score'],inplace = True)

In [11]:
print(df_titles['genres'].isnull().sum())
print(df_titles['imdb_score'].isnull().sum())

0
0


In [12]:
df_titles['release_year'] = pd.to_datetime(df_titles['release_year'])

In [13]:
df_titles['production_countries'] = df_titles['production_countries'].astype('category')

发现`seasons`中存在缺失值，所以无法顺利转换成int类型。由于`seasons`并非本次数据分析的重要变量，故可以忽略。

由于前面我们已经对`genres`列里的值拆分成多个独立的行，所以`df_titles`中大部分的变量都是可以重复的，无需删除。

但是如果有两行所有的变量都重复，那么就属于录入错误，可以删除。

In [14]:
df_titles.duplicated().sum()

np.int64(0)

发现没有所有变量都重复的行，故可以省略删除操作。

查看重要变量`imdb_score`有没有无效/错误数据

In [15]:
df_titles.describe()

release_year       runtime      seasons    imdb_score  \
count                          14393  14393.000000  5655.000000  14393.000000   
mean   1970-01-01 00:00:00.000002015     78.326270     2.424934      6.534802   
min    1970-01-01 00:00:00.000001954      0.000000     1.000000      1.500000   
25%    1970-01-01 00:00:00.000002015     44.000000     1.000000      5.800000   
50%    1970-01-01 00:00:00.000002018     87.000000     1.000000      6.600000   
75%    1970-01-01 00:00:00.000002020    106.000000     3.000000      7.400000   
max    1970-01-01 00:00:00.000002022    225.000000    42.000000      9.500000   
std                              NaN     40.555015     2.867856      1.157513   

         imdb_votes  tmdb_popularity    tmdb_score  
count  1.437100e+04     14268.000000  13975.000000  
mean   2.836162e+04        30.008057      6.885143  
min    5.000000e+00         0.600000      1.000000  
25%    7.080000e+02         3.911750      6.200000  
50%    3.166000e+03         9.666000      6.985000  
75%    1.432100e+04        22.766000      7.600000  
max    2.294231e+06      2274.044000     10.000000  
std    1.041849e+05        99.359539      1.112448

#### df_credits

In [16]:
df_credits.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77801 entries, 0 to 77800
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   person_id  77801 non-null  int64 
 1   id         77801 non-null  object
 2   name       77801 non-null  object
 3   character  68029 non-null  object
 4   role       77801 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.0+ MB


该数据集中有77801行观察值：

缺失值：`character`有缺失值。

数据类型错误：`person_id`应转为str，但是该变量数据类型正确与否不影响本次数据分析，所以无需处理。

因为本次数据分析的目的是，寻找出各个流派中平均作品评分最高的演员，所以`character`是重要的变量，如若缺失，应删除。

In [17]:
df_credits = df_credits.dropna(subset = ['character'])

In [18]:
df_credits['character'].isnull().sum()

np.int64(0)

已经将所有`character`变量为空值的行删除。

In [19]:
df_credits

person_id         id                name                character  \
0           3748    tm84618      Robert De Niro            Travis Bickle   
1          14658    tm84618        Jodie Foster            Iris Steensma   
2           7064    tm84618       Albert Brooks                      Tom   
3           3739    tm84618       Harvey Keitel  Matthew 'Sport' Higgins   
4          48933    tm84618     Cybill Shepherd                    Betsy   
...          ...        ...                 ...                      ...   
77795     368473  tm1059008        Aída Morales                  Maritza   
77796     736339  tm1059008    Adelaida Buscato                María Paz   
77797     399499  tm1059008  Luz Stella Luengas             Karen Bayona   
77798     373198  tm1059008         Inés Prieto                    Fanny   
77799     378132  tm1059008        Isabel Gaona                   Cacica   

        role  
0      ACTOR  
1      ACTOR  
2      ACTOR  
3      ACTOR  
4      ACTOR  
...      ...  
77795  ACTOR  
77796  ACTOR  
77797  ACTOR  
77798  ACTOR  
77799  ACTOR  

[68029 rows x 5 columns]

假若`df_credits`中有观察值所有的变量都重复，那么就属于录入错误，可以删除。

In [20]:
df_credits.duplicated().sum()

np.int64(0)

确认`df_credits`没有所有变量都重复的观察存在。

由于`df_credits`中没有重要的数值类型的变量，所以无需查看无效/错误数据。

## 整理数据

本次数据分析的目的是，查找各个流派中作品评分最高的演员。

这既需要`genres`&`imdb_scores`变量，也需要`person_ID`&`role`变量，因此需要使用`merge`合并两个表格。

在调用`merge`方法之前，需要先把`df_credits`的`role`变量中非`ACTOR`值的观察值删除。

In [21]:
df_credits = df_credits.query('role == "ACTOR"')

In [22]:
df_credits.query('role != "ACTOR"')

Empty DataFrame
Columns: [person_id, id, name, character, role]
Index: []

经过调用`query`函数并重新赋值给`df_credits`，成功删除`df_credits`的`role`变量中非`ACTOR`值的观察值。

In [23]:
titles_with_credits = pd.merge(df_titles,df_credits,on = 'id', how = 'inner')
titles_with_credits

id        title   type  \
0         tm84618  Taxi Driver  MOVIE   
1         tm84618  Taxi Driver  MOVIE   
2         tm84618  Taxi Driver  MOVIE   
3         tm84618  Taxi Driver  MOVIE   
4         tm84618  Taxi Driver  MOVIE   
...           ...          ...    ...   
194281  tm1059008      Lokillo  MOVIE   
194282  tm1059008      Lokillo  MOVIE   
194283  tm1059008      Lokillo  MOVIE   
194284  tm1059008      Lokillo  MOVIE   
194285  tm1059008      Lokillo  MOVIE   

                                              description  \
0       A mentally unstable Vietnam War veteran works ...   
1       A mentally unstable Vietnam War veteran works ...   
2       A mentally unstable Vietnam War veteran works ...   
3       A mentally unstable Vietnam War veteran works ...   
4       A mentally unstable Vietnam War veteran works ...   
...                                                   ...   
194281  A controversial TV host and comedian who has b...   
194282  A controversial TV host and comedian who has b...   
194283  A controversial TV host and comedian who has b...   
194284  A controversial TV host and comedian who has b...   
194285  A controversial TV host and comedian who has b...   

                        release_year age_certification  runtime  genres  \
0      1970-01-01 00:00:00.000001976                 R      114   drama   
1      1970-01-01 00:00:00.000001976                 R      114   drama   
2      1970-01-01 00:00:00.000001976                 R      114   drama   
3      1970-01-01 00:00:00.000001976                 R      114   drama   
4      1970-01-01 00:00:00.000001976                 R      114   drama   
...                              ...               ...      ...     ...   
194281 1970-01-01 00:00:00.000002021               NaN       90  comedy   
194282 1970-01-01 00:00:00.000002021               NaN       90  comedy   
194283 1970-01-01 00:00:00.000002021               NaN       90  comedy   
194284 1970-01-01 00:00:00.000002021               NaN       90  comedy   
194285 1970-01-01 00:00:00.000002021               NaN       90  comedy   

       production_countries  seasons     imdb_id  imdb_score  imdb_votes  \
0                    ['US']      NaN   tt0075314         8.2    808582.0   
1                    ['US']      NaN   tt0075314         8.2    808582.0   
2                    ['US']      NaN   tt0075314         8.2    808582.0   
3                    ['US']      NaN   tt0075314         8.2    808582.0   
4                    ['US']      NaN   tt0075314         8.2    808582.0   
...                     ...      ...         ...         ...         ...   
194281               ['CO']      NaN  tt14585902         3.8        68.0   
194282               ['CO']      NaN  tt14585902         3.8        68.0   
194283               ['CO']      NaN  tt14585902         3.8        68.0   
194284               ['CO']      NaN  tt14585902         3.8        68.0   
194285               ['CO']      NaN  tt14585902         3.8        68.0   

        tmdb_popularity  tmdb_score  person_id                name  \
0                40.965       8.179       3748      Robert De Niro   
1                40.965       8.179      14658        Jodie Foster   
2                40.965       8.179       7064       Albert Brooks   
3                40.965       8.179       3739       Harvey Keitel   
4                40.965       8.179      48933     Cybill Shepherd   
...                 ...         ...        ...                 ...   
194281           26.005       6.300     368473        Aída Morales   
194282           26.005       6.300     736339    Adelaida Buscato   
194283           26.005       6.300     399499  Luz Stella Luengas   
194284           26.005       6.300     373198         Inés Prieto   
194285           26.005       6.300     378132        Isabel Gaona   

                      character   role  
0                 Travis Bickle  ACTOR  
1                 Iris Steensma  ACTOR  
2          

使用`groupby`方法，根据`genres`&`person_id`分组，求`imdb_score`的平均值。

In [24]:
imdb_score_mean = titles_with_credits.groupby(['genres','person_id'])['imdb_score'].mean()
imdb_score_mean

genres   person_id
action   45           5.0
         48           5.4
         51           6.4
         53           6.8
         54           5.3
                     ... 
western  2353339      6.9
         2370848      6.1
         2398539      3.8
         2406218      6.0
         2408082      7.3
Name: imdb_score, Length: 160308, dtype: float64

为了后续操作方便，将该多层索引df转为正常的df.

In [25]:
imdb_score_mean_reset = imdb_score_mean.reset_index()
imdb_score_mean_reset

genres  person_id  imdb_score
0        action         45         5.0
1        action         48         5.4
2        action         51         6.4
3        action         53         6.8
4        action         54         5.3
...         ...        ...         ...
160303  western    2353339         6.9
160304  western    2370848         6.1
160305  western    2398539         3.8
160306  western    2406218         6.0
160307  western    2408082         7.3

[160308 rows x 3 columns]

为了能更直观的查看高评分演员，应使用`merge`方法，把`name`的值加入进来。

在加入`name`之前，需要先删除重复值，避免合并后产生冗余数据。

In [26]:
person_id_with_name = df_credits[['person_id','name']]
person_id_with_name

person_id                name
0           3748      Robert De Niro
1          14658        Jodie Foster
2           7064       Albert Brooks
3           3739       Harvey Keitel
4          48933     Cybill Shepherd
...          ...                 ...
77795     368473        Aída Morales
77796     736339    Adelaida Buscato
77797     399499  Luz Stella Luengas
77798     373198         Inés Prieto
77799     378132        Isabel Gaona

[68029 rows x 2 columns]

In [27]:
person_id_with_name = person_id_with_name.drop_duplicates()

In [28]:
person_id_with_name.duplicated().sum()

np.int64(0)

In [29]:
imdb_score_with_name = pd.merge(imdb_score_mean_reset,person_id_with_name,on = 'person_id', how = 'inner')
imdb_score_with_name

genres  person_id  imdb_score                 name
0        action         45         5.0          David Lewis
1        action         48         5.4             Tom Kane
2        action         51         6.4         Kevin Conroy
3        action         53         6.8          Phil LaMarr
4        action         54         5.3      Susan Eisenberg
...         ...        ...         ...                  ...
160303  western    2353339         6.9         Pascal Nowak
160304  western    2370848         6.1           Suraj Vyas
160305  western    2398539         3.8  Ashley Luren Nichol
160306  western    2406218         6.0                Jumpy
160307  western    2408082         7.3     Jesse Youngblood

[160308 rows x 4 columns]

In [36]:
genres_max_score = imdb_score_with_name.groupby('genres')['imdb_score'].max()
genres_max_score

genres
action           9.3
animation        9.3
comedy           9.2
crime            9.5
documentation    9.1
drama            9.5
european         8.9
family           9.3
fantasy          9.3
history          9.1
horror           9.0
music            8.8
reality          8.7
romance          9.2
scifi            9.3
sport            9.1
thriller         9.5
war              8.8
western          8.9
Name: imdb_score, dtype: float64

In [38]:
genres_max_score_with_name = pd.merge(genres_max_score,imdb_score_with_name,on=['genres','imdb_score'])
genres_max_score_with_name

genres  imdb_score  person_id                name
0     action         9.3       1303       Jessie Flower
1     action         9.3      12790         Olivia Hack
2     action         9.3      21033          Zach Tyler
3     action         9.3      86591       Cricket Leigh
4     action         9.3     336830     André Sogliuzzo
..       ...         ...        ...                 ...
131      war         8.8     826547         Yuto Uemura
132  western         8.9      22311     Koichi Yamadera
133  western         8.9      28166  Megumi Hayashibara
134  western         8.9      28180      Unsho Ishizuka
135  western         8.9      93017            Aoi Tada

[136 rows x 4 columns]

In [39]:
genres_max_score_with_name.to_csv('genres_max_score_with_name.csv',index=False)